### PR TITLE
native: introduce time scaling

### PR DESCRIPTION
**Pull request by miri64**
One step more into the direction of using native for simulation.

### CODE REVIEW

**Comment by miri64** [comment]
(in case this is not clear, this feature is very experimental ;-))

File: cpu/native/periph/timer.c [hunk 1]
```diff
@@ -62,7 +65,7 @@ static struct itimerval itv;
 static unsigned long ts2ticks(struct timespec *tp)
 {
     /* TODO: check for overflow */
-    return((tp->tv_sec * NATIVE_TIMER_SPEED) + (tp->tv_nsec / 1000));
+    return (((tp->tv_sec * NATIVE_TIMER_SPEED) + (tp->tv_nsec / 1000)) << native_timer_scale);
```
Comment (kaspar030) [inline cpu/native/periph/timer.c]: Do we have to worry for overflows here?

File: cpu/native/periph/timer.c [hunk 2]
```diff
@@ -109,6 +112,7 @@ static void do_timer_set(unsigned int offset)
         offset = NATIVE_TIMER_MIN_RES;
     }
 
+    offset >>= native_timer_scale;
```
Comment (kaspar030) [inline cpu/native/periph/timer.c]: ... or underflows?

File: cpu/native/periph/timer.c [hunk 3]
```diff
@@ -47,7 +47,10 @@
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
-#define NATIVE_TIMER_SPEED 1000000
+#define NATIVE_TIMER_SPEED          (1000000)
+#define NATIVE_TIMER_DEFAULT_SCALE  (0)
```
Comment (kaspar030) [inline cpu/native/periph/timer.c]: Please add a sentence or two explaining what ```native_timer_scale``` is used for.

File: cpu/native/startup.c [hunk 4]
```diff
@@ -217,6 +220,8 @@ void usage_exit(int status)
 "    -s <seed>, --seed=<seed>\n"
 "        specify srandom(3) seed (/dev/random is used instead of random(3) if\n"
 "        the option is omitted)\n"
+"    -t <x>, --timer-scale=<x>\n"
+"        Scaling (speed-up or slow-down) of the timers by factor 2^x (default = 0)\n"
```
Comment (kaspar030) [inline cpu/native/startup.c]: Does it speed up or slow down?

Comment (miri64) [inline cpu/native/startup.c]: Depends on the signage. `2^x < 1` if `x < 0` ;)

File: cpu/native/periph/timer.c (hunk 3 above)
Comment (miri64) [inline cpu/native/periph/timer.c]: Will do.

File: cpu/native/periph/timer.c (hunk 1 above)
Comment (miri64) [inline cpu/native/periph/timer.c]: Yes, but that was the case before (see comment above)

File: cpu/native/periph/timer.c (hunk 2 above)
Comment (miri64) [inline cpu/native/periph/timer.c]: Yes, I think so.

**Comment by kaspar030** [comment]
> Depends on the signage. |2^x < 1| if |x < 0| ;)

Not sure that'll work as expected:

```
[kaspar@booze src]$ cat test.c
#include <stdio.h>

int main(void)
{
    int n = -1;
    printf("%u\n", 16384 >> n);
    return 0;
}

[kaspar@booze src]$ gcc -Wall -pedantic test.c -o test
[kaspar@booze src]$ ./test
0
[kaspar@booze src]$
```

File: cpu/native/startup.c (hunk 4 above)
Comment (miri64) [inline cpu/native/startup.c]: > Not sure that'll work as expected:

Indeed, I expected it to just shift the other way. Will fix.

**Comment by miri64** [comment]
Rebased and addressed comments.

File: cpu/native/periph/timer.c (hunk 3 above)
Comment (miri64) [reply cpu/native/periph/timer.c]: Done

File: cpu/native/startup.c (hunk 4 above)
Comment (miri64) [reply cpu/native/startup.c]: Done

**Comment by miri64** [comment]
Ping @kaspar030?

**Comment by miri64** [comment]
Ping @kaspar030?

File: cpu/native/periph/timer.c [hunk 5]
```diff
@@ -62,7 +65,14 @@ static struct itimerval itv;
 static unsigned long ts2ticks(struct timespec *tp)
 {
     /* TODO: check for overflow */
-    return((tp->tv_sec * NATIVE_TIMER_SPEED) + (tp->tv_nsec / 1000));
+    if (native_timer_scale < 0) {
```
Comment (kaspar030) [inline cpu/native/periph/timer.c]: Maybe dedup as much as possible:

```
unsigned long result = (tp->tv_sec * NATIVE_TIMER_SPEED) + (tp->tv_nsec / 1000);
if (native_timer_scale >= 0) {
  result <<= native_timer_scale;
} else {
  result >>= -native_timer_scale;
}
return result;
```

Comment (miri64) [reply cpu/native/periph/timer.c]: The `>=` you just put in to mess with my head ;-)?

Comment (miri64) [reply cpu/native/periph/timer.c]: Done.

**Comment by miri64** [comment]
I also moved the `extern unsigned int native_timer_scale` in `startup.c` to `extern int native_timer_scale`, because obviously `native_timer_scale` can be negative ;-).

**Comment by miri64** [comment]
Ping @kaspar030?

**Comment by kaspar030** [comment]
Please squash!

Ffeel free to factor out sth like this:

```
static unsigned long signed_shift(unsigned long val, int shift)
{
  if (shift > 0) {
    val <<= shift;
} else if (shift < 0) {
  val >>= -shift;
}
return val;
}
```

**Comment by miri64** [comment]
Squashed and factored out the function.

**Comment by kYc0o** [comment]
@miri64 can you address Murdock errors?

**Comment by miri64** [comment]
Rebased and adapted to master (which should fix the Murdock issues).

**Comment by miri64** [comment]
Rebased and squashed. Ping @kaspar030?

File: cpu/native/periph/timer.c [hunk 6]
```diff
@@ -113,6 +129,13 @@ static void do_timer_set(unsigned int offset)
         offset = NATIVE_TIMER_MIN_RES;
     }
 
+    if (native_timer_scale < 0) {
```
Comment (kaspar030) [inline cpu/native/periph/timer.c]: shouldn't this use ```signed_shift(offset, -native_timer_scale)```?

**Pull request closed without merging**